RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronSeedProducer.cc

```diff
@@ -0,0 +1,536 @@
+  particle.setCharge(kfTrackRef->charge());
+  particle.propagateToEcalEntrance(false);
+  if ( particle.getSuccess() == 0 ) { return; }
+  
+  // ECAL entry point for track
+  GlobalPoint ecal_pos(particle.vertex().x(),
+		       particle.vertex().y(),
+		       particle.vertex().z());
+
+  // Preshower limit
+  bool below_ps = pow(ecal_pos.z(),2.) > boundary_*ecal_pos.perp2();
+  
+  // Iterate through ECAL clusters 
+  for ( unsigned int iclu = 0; iclu < ecalClusters.product()->size(); iclu++ ) {
+    reco::PFClusterRef cluRef(ecalClusters,iclu);
+
+    // Correct ecal_pos for shower depth 
+    double shower_depth = reco::PFCluster::getDepthCorrection(cluRef->correctedEnergy(),
+							      below_ps,
+							      false);
+    GlobalPoint showerPos = ecal_pos + 
+      GlobalVector(particle.momentum().x(),
+		   particle.momentum().y(),
+		   particle.momentum().z()).unit() * shower_depth;
+
+    // Determine dR squared
+    float dr2 = reco::deltaR2( cluRef->positionREP(), showerPos );
+
+    // Find nearest ECAL cluster
+    if ( dr2 < info.dr2min ) {
+      info.dr2min = dr2;
+      info.cluRef = cluRef;
+      info.deta = std::abs( cluRef->positionREP().eta() - showerPos.eta() );
+      info.dphi = 
+	std::abs( reco::deltaPhi( cluRef->positionREP().phi(), showerPos.phi() )) * 
+	kfTrackRef->charge();
+      info.showerPos = showerPos;
+    }
+  
+  }
+
+  // Populate PreId object
+  math::XYZPoint point( ecal_pos.x(),
+			ecal_pos.y(),
+			ecal_pos.z() );
+
+  // Set PreId content
+  ecalPreId.setECALMatchingProperties( info.cluRef,
+				       point,
+				       info.showerPos,
+				       info.deta,
+				       info.dphi,
+				       0.f, // chieta
+				       0.f, // chiphi
+				       kfTrackRef->normalizedChi2(), // chi2
+				       info.cluRef->correctedEnergy() / std::sqrt( kfTrackRef->innerMomentum().mag2() ) ); // E/p
+
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// Original implementation for "lightweight" GSF tracking
+bool LowPtGsfElectronSeedProducer::lightGsfTracking( reco::PreId& preId,
+						     const reco::TrackRef& trackRef,
+						     const reco::ElectronSeed& seed,
+						     const edm::EventSetup& setup )
+{
+  
+  edm::ESHandle<TrajectoryFitter> fitter;
+  setup.get<TrajectoryFitter::Record>().get(fitter_,fitter);
+  std::unique_ptr<TrajectoryFitter> fitterPtr = fitter->clone();
+
+  edm::ESHandle<TrajectorySmoother> smoother;
+  setup.get<TrajectoryFitter::Record>().get(smoother_,smoother);
+  std::unique_ptr<TrajectorySmoother> smootherPtr;
+  smootherPtr.reset(smoother->clone());
+
+  edm::ESHandle<TransientTrackingRecHitBuilder> builder;
+  setup.get<TransientRecHitRecord>().get(builder_,builder);
+  TkClonerImpl hitCloner = static_cast<TkTransientTrackingRecHitBuilder const*>(builder.product())->cloner();
+  fitterPtr->setHitCloner(&hitCloner);
+  smootherPtr->setHitCloner(&hitCloner);
```

Comment:
@perrotta 
related to your notes in https://github.com/cms-sw/cmssw/pull/25455#issuecomment-449980257
```almost all the cpu is taken by the loop() method, and inside it 78.5% of the time is spent by lightGsfTracking()```.
Please check if the cost of  ```fitter->clone()``` and ```smoother->clone()``` as well as their destructors is minimal.
I expect that all these ES tools can be extracted once per event instead of once per track.
Also, shouldn't there already be a way to get both the fitter and the smoother from the event setup with a correct hit cloner.

Reply:
@slava77 
Moving the definitions and those two clonings at the beginning of the produce method was indeed the first (and only, so far) attempt I made to speed up the loop. I did not find any improvement, unfortunately: the cpu spent in lightGsfTracking() was in fact basically the same as before (within the uncertainties  of the method)

Reply:
Hi @perrotta, @slava77, many thanks for following up on the timing studies. 

Are these the relevant numbers:
```=== Loose vs Orig: Job total:  8.51035 s/ev ==> 9.44562 s/ev```
```=== Tight vs Orig: Job total:  8.51035 s/ev ==> 9.12637 s/ev```

If so, I'm not sure where you get the following percentage increases: "Overall, this PR increases the overal Reco processing time by 9.1% with the Tight WP, and by 12,4% with the Loose WP." Please can you clarify? 

In terms of raising the pT threshold of tracks considered by the ElectronSeed producer, as suggested by @slava77 in [this](https://github.com/cms-sw/cmssw/pull/25455#discussion_r245066660) comment, I am investigating this now. 

In terms of the igprof output and code optimisation, it's not obvious to me where to start within the ```lightGsfTracking``` method. Let's see what the change in pT threshold brings ...

Reply:
Ok, here we go ...

I'm using the 25202 TTbar RelVal sample.
I'm always using the ```Tight``` working point, as is currently default. 
I raise the pT threshold of the tracks considered by the ElectronSeed producer, [here](https://github.com/bainbrid/cmssw/blob/LowPtElectrons_10X/RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronSeeds_cfi.py#L31).
Three thresholds are considered, pT > 0.5 (current default), 0.7, and 1.0 GeV.
The effect of the timing is considered, relative to the "nominal" reco, _without_ low pT GsfTracks.

In summary, the timing increases by 7.7, 5.5, and 3.5% for the thresholds 0.5, 0.7, and 1.0 GeV, respectively. 

The 0.7 GeV threshold corresponds to the minimum pT required for electrons to reach the barrel ECAL. The 1.0 GeV threshold starts to eat into our B->Kee signal acceptance, mainly because the subheading electron pT distribution peaks at ~1 GeV. The (GEN-level) acceptance for the full "Kee" final state reduces by 16% (0.5->0.7 GeV) and 26% (0.5->1.0 GeV). 

```
pT > 0.5 GeV:
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +7.25%         0.00 ms/ev ->       639.28 ms/ev lowPtGsfElectronSeeds
       added      +0.32%         0.00 ms/ev ->        28.50 ms/ev lowPtGsfEleGsfTracks
       added      +0.05%         0.00 ms/ev ->         4.66 ms/ev lowPtGsfEleCkfTrackCandidates
       added      +0.08%         0.00 ms/ev ->         7.15 ms/ev lowPtGsfElePfTracks
  ---------- ------------     --------                  ----       ------------
       TOTAL	  +7.70%
Job total:  8.81418 s/ev ==> 9.74298 s/ev 

pT > 0.7 GeV:
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +5.04%         0.00 ms/ev ->       444.03 ms/ev lowPtGsfElectronSeeds
       added      +0.33%         0.00 ms/ev ->        29.35 ms/ev lowPtGsfEleGsfTracks
       added      +0.06%         0.00 ms/ev ->         4.89 ms/ev lowPtGsfEleCkfTrackCandidates
       added      +0.08%         0.00 ms/ev ->         7.16 ms/ev lowPtGsfElePfTracks
  ---------- ------------     --------                  ----       ------------
       TOTAL	  +5.51%
Job total:  8.81418 s/ev ==> 9.57468 s/ev

pT > 1.0 GeV:
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +3.07%         0.00 ms/ev ->       270.56 ms/ev lowPtGsfElectronSeeds
       added      +0.33%         0.00 ms/ev ->        29.05 ms/ev lowPtGsfEleGsfTracks
       added      +0.05%         0.00 ms/ev ->         4.62 ms/ev lowPtGsfEleCkfTrackCandidates
       added      +0.08%         0.00 ms/ev ->         7.09 ms/ev lowPtGsfElePfTracks
  ---------- ------------     --------                  ----       ------------
       TOTAL	  +3.53%
Job total:  8.81418 s/ev ==> 9.28047 s/ev
```

Reply:
That looks promising, in particular for the default value to be kept for the standard workflows. For the configuration to be run to process the 2018 parked data I would stick on the more expensive but also more efficient one. But we can discuss it further during the reco meeting on Friday.

In parallel, could you please move the  two ```fitter->clone()``` and ```smoother->clone()``` and definitions at the beginning of the produce() method, or whenever else it can be convenient to have them applied only once, and not for every track? Even if I did not notice any significant improvement in timing, this is a more logical place where to apply them.

Reply:
Concerning moving the fitter and smoother, this is done, [here](
https://github.com/cms-sw/cmssw/pull/25455/commits/318b6cee2d6359a5eae141dd352c211842c76887)

Reply:
@slava77 @perrotta Concerning the timing, I've raised the default pT threshold from 0.5 to 1.0 GeV for the "standard workflows", such that the timing increase is only 3.5%. For the bParking era, we override the value and our preferred value of 0.5 GeV. The changes can be found [here](https://github.com/cms-sw/cmssw/pull/25455/commits/593f32261d270d4be30c79fe6cac1d58352abce5).